receiver/prometheusreceiver/internal/metrics_adjuster.go

```diff
@@ -0,0 +1,169 @@
+package internal
+
+import (
+	"fmt"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"go.uber.org/zap"
+	"strings"
+)
+
+// MetricsAdjuster takes a map from a metric instance to the initial point in the metrics instance
+// and provides AdjustMetrics, which takes a sequence of metrics and adjust their values based on
+// the initial points.
+type MetricsAdjuster struct {
+	metricMap *map[string]*metricspb.TimeSeries
+	logger    *zap.SugaredLogger
+}
+
+// NewMetricsAdjuster is a constructor for MetricsAdjuster.
+func NewMetricsAdjuster(metricMap *map[string]*metricspb.TimeSeries, logger *zap.SugaredLogger) *MetricsAdjuster {
+	return &MetricsAdjuster{
+		metricMap: metricMap,
+		logger:    logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their values based on the initial points in the
+// metricMap. If the metric is the first point in the timeseries, or the timeseries has been reset, it is
+// removed from the sequence and added to the the metricMap.
+func (ma *MetricsAdjuster) AdjustMetrics(metrics []*metricspb.Metric) []*metricspb.Metric {
+	var adjusted = make([]*metricspb.Metric, 0, len(metrics))
+	for _, metric := range metrics {
+		if ma.adjustMetric(metric) {
+			adjusted = append(adjusted, metric)
+		}
+	}
+	return adjusted
+}
+
+func (ma *MetricsAdjuster) adjustMetric(metric *metricspb.Metric) bool {
+	switch metric.MetricDescriptor.Type {
+	case metricspb.MetricDescriptor_GAUGE_DOUBLE:
+		return true
+	default:
+		return ma.adjustMetricTimeseries(metric)
+	}
+}
+
+func (ma *MetricsAdjuster) adjustMetricTimeseries(metric *metricspb.Metric) bool {
+	filtered := make([]*metricspb.TimeSeries, 0, len(metric.GetTimeseries()))
+	for _, timeseries := range metric.GetTimeseries() {
+		key := getSignature(metric.GetMetricDescriptor().GetName(), timeseries.GetLabelValues())
+		initial, ok := (*ma.metricMap)[key]
+		if ok {
+			if !ma.adjustTimeseries(metric.MetricDescriptor.Type, timeseries, initial) {
+				// detected a reset
+				(*ma.metricMap)[key] = timeseries
+			}
+			filtered = append(filtered, timeseries)
+		} else {
+			(*ma.metricMap)[key] = timeseries
+		}
+	}
+	if len(filtered) == 0 {
+		return false
+	}
+	metric.Timeseries = filtered
+	return true
+}
+
+func (ma *MetricsAdjuster) adjustTimeseries(metricType metricspb.MetricDescriptor_Type, current, initial *metricspb.TimeSeries) bool {
+	if !ma.adjustPoints(metricType, current.GetPoints(), initial.GetPoints()) {
+		return false
+	}
+	current.StartTimestamp = initial.StartTimestamp
+	return true
+}
+
+func (ma *MetricsAdjuster) adjustPoints(metricType metricspb.MetricDescriptor_Type, current, initial []*metricspb.Point) bool {
```

Comment:
you can probably get rid of adjustPoints.  In adjustTimeseries
```go
func (ma *MetricsAdjuster) adjustTimeseries(metricType metricspb.MetricDescriptor_Type, current, initial *metricspb.TimeSeries) bool {
        cPoints := current.GetPoints()
        iPoints := initial.GetPoints()
        if len(cPoints) !=1 || len(iPoints) != 1 {
            // log if needed.
            return false
        }

	if !ma.adjustPoint(metricType, cPoints[0], iPoints[0]) {
		return false
	}
	current.StartTimestamp = initial.StartTimestamp
	return true
}
```

Reply:
I generally avoid using Error unless something major is going on but that's based on Google prod convention - what is the convention for the oc agent when Error is appropriate?

Reply:
I've simplified it - ptal.